azkaban-common/src/main/java/azkaban/version/JdbcVersionSetLoaderImpl.java

```diff
@@ -0,0 +1,217 @@
+package azkaban.version;
+
+import azkaban.db.DatabaseOperator;
+import azkaban.db.SQLTransaction;
+import java.io.IOException;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.commons.dbutils.ResultSetHandler;
+import org.apache.log4j.Logger;
+
+/**
+ * A JDBC based implementation for loading VersionSet from the backed db.
+ * This class keeps the copy of entries in the version_set table locally
+ * as well to reduce the number of queries to the db.
+ *
+ * This is a singleton class to be instantiated using dependency injection.
+ */
+@Singleton
+public class JdbcVersionSetLoaderImpl implements VersionSetLoader {
+    private static final Logger logger = Logger.getLogger(JdbcVersionSetLoaderImpl.class);
+
+    private static final String SELECT_VSET_FROM_NUM = "SELECT vset.num, vset.md5, vset.json FROM version_set vset WHERE vset.num=?";
+    private static final String SELECT_VSET_FROM_MD5 = "SELECT vset.num, vset.md5, vset.json FROM version_set vset WHERE vset.md5=?";
+    private static final String SELECT_ALL_VSET = "SELECT vset.num, vset.md5, vset.json FROM version_set vset";
+    private static final String INSERT_VSET = "INSERT INTO version_set (md5, json) values (?,?)";
+    private static final String DELETE_VSET = "DELETE FROM version_set vset WHERE vset.md5=?";
+
+    private final DatabaseOperator dbOperator;
+    private final Map<String, VersionSet> md5ToVersionSet = new HashMap<>();
```

Comment:
If the user provides versionSetId/versionSetNum as part of flow parameters and we want to determine the version set corresponding to it, then we need this mapping (otherwise we may have to iterate over all values and find it). Also, the versionSet object itself is common for both the maps.

Even with LRU, we may have to end up either using two caches or maintaining one cache and one map and keeping them in sync. Also, it may be overkill in the beginning. I think we should see the size of the version set table and then update the code to the LRU cache accordingly.